### PR TITLE
Fix handling unreachability in query action

### LINKED ISSUE #34744: Unreachability not checked within query actions
**Description:**
```ballerina
check from var item in 1...5
where false
do {
    io:println("this is unreachable");
};
```

**Steps to reproduce:**

**Affected Versions:**
SL GA RC1

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #34744

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java [hunk 1]
```diff
@@ -184,14 +216,13 @@ public void visit(BLangSimpleVariableDef varDefNode) {
...
-        if (expr != null && expr.getKind() == NodeKind.DO_ACTION) {
-            analyzeReachability(expr, env);
-        }
+        analyzeReachabilityInExpressionIfApplicable(expr, env);
     }
 
     @Override
     public void visit(BLangAssignment assignNode) {
         checkStatementExecutionValidity(assignNode);
+        analyzeReachabilityInExpressionIfApplicable(assignNode.expr, env);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java:225]: Need to handle destructuring assignment statements also. The following doesn't result in an error atm.

```ballerina
public function main() returns error? {
    var error(m) = <error> from var item in 1 ... 5
        where false
        do {
            int _ = 1;
        };
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java [hunk 2]
```diff
@@ -628,7 +672,9 @@ public void visit(BLangDoClause doClause) {
                                                                                               env.enclInvokable));
 
         BLangBlockStmt body = doClause.body;
-        checkStatementExecutionValidity(body);
+        if (!(doClause.parent != null && doClause.parent.getKind() == NodeKind.DO_ACTION)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java:675]: Wouldn't it be clearer to say

```suggestion
        if (doClause.parent == null || doClause.parent.getKind() != NodeKind.DO_ACTION) {
```

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java:675]: Why do we need this check in the first place?

File: tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/unreachability_test.bal [hunk 3]
```diff
@@ -1043,3 +1043,187 @@ function testUnreachableStatementInQueryAction() {
 
     _ = n is error; // reachable
 }
+
+function testUnreachableStatementInQueryAction2() returns error? {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/unreachability_test.bal:1047]: Shall we add some (positive) tests with `where true` also?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java (hunk 1 above)
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java:225]: Fixed

**Review by MaryamZi — DISMISSED** [review]

**Review by MaryamZi — APPROVED** [review]

**Review by pcnfernando — APPROVED** [review]
Tested

**Pull request closed without merging**